### DISCUSSION TITLE
What should be the callback when using GitHub Actions?

### DISCUSSION THREAD

**Discussion by arnonrdp** [tweepy/tweepy#2142] [Questions] [answered]
### Question

My application runs through a GitHub Action.What should my callback be?
I have a cron job that makes my app post some news at a certain time of day.
But after this Twitter update to v2, it never worked again.

Então, atualizei aquele antigo modelo:
```py
# OLD - It worked in v1
auth = OAuthHandler(getenv('API_KEY'),
                    getenv('API_KEY_SECRET'))
auth.set_access_token(getenv('ACCESS_TOKEN'),
                      getenv('ACCESS_TOKEN_SECRET'))
api = API(auth)
api.verify_credentials()
```

But I can't get this to work.
Looks like it got more complex.

### Relevant Code

```python
def twitter_connect(tweet_list):
    auth = OAuth2UserHandler(client_id=getenv('CLIENT_ID'),
                             redirect_uri='https://github.com/arnonrdp/Bot-Deschamps-Newsletter',
                             scope=['tweet.write'],
                             client_secret=getenv('API_KEY_SECRET'))
    url = auth.get_authorization_url()
    access_token = auth.fetch_token(url)
    api = Client(access_token)

    try:
        api.verify_credentials()
        print('Good news, you are logged!')
        prepare_tweet(tweet_list, api)
    except Exception as e:
        print('Bad news, will try again in 2 minutes.\n', e)
        sleep(120)
        twitter_connect(tweet_list)
```

### Traceback

```python-traceback
Traceback (most recent call last):
  File "/Users/arnon/Public/GitHub/Bot-Deschamps-Newsletter/post_tweet.py", line 15, in twitter_connect
    access_token = auth.fetch_token(url)
  File "/Users/arnon/Public/GitHub/Bot-Deschamps-Newsletter/.venv/lib/python3.9/site-packages/tweepy/auth.py", line 217, in fetch_token
    return super().fetch_token(
  File "/Users/arnon/Public/GitHub/Bot-Deschamps-Newsletter/.venv/lib/python3.9/site-packages/requests_oauthlib/oauth2_session.py", line 244, in fetch_token
    self._client.parse_request_uri_response(
  File "/Users/arnon/Public/GitHub/Bot-Deschamps-Newsletter/.venv/lib/python3.9/site-packages/oauthlib/oauth2/rfc6749/clients/web_application.py", line 220, in parse_request_uri_response
    response = parse_authorization_code_response(uri, state=state)
  File "/Users/arnon/Public/GitHub/Bot-Deschamps-Newsletter/.venv/lib/python3.9/site-packages/oauthlib/oauth2/rfc6749/parameters.py", line 284, in parse_authorization_code_response
    raise MissingCodeError("Missing code parameter in response.")
oauthlib.oauth2.rfc6749.errors.MissingCodeError: (missing_code) Missing code parameter in response.
Auth:
 https://twitter.com/i/oauth2/authorize?response_type=code&client_id=dUNFZEN5dVZVZGItcmZKWWFYQXQ6MTpjaQ&redirect_uri=https%3A%2F%2Fgithub.com%2Farnonrdp%2FBot-Deschamps-Newsletter&scope=tweet.write&state=cvGy54WBy4pYUrrjQW7jzPxIFFJVCL&code_challenge=CTcbeAR3NkSBBiLLGZ-l1lqsjRnZW35Wy0qm7_oO2pg&code_challenge_method=S256
```

### Twitter API Access Plan

Free

### Tweepy Version

4.14.0

### Checklist

- [X] I have searched for duplicate questions.
- [X] If applicable, I have removed any visible credentials from my code and/or screenshots.

**Comment by Harmon758** [comment]
You need to actually use the authorization URL, instead of just passing it to `OAuth2UserHandler.fetch_token`.
See https://tweepy.readthedocs.io/en/v4.14.0/authentication.html#oauth-2-0-authorization-code-flow-with-pkce-user-context.

You can continue using OAuth 1.0a User Context with Twitter API v2 though.

Also note, `Client` does not have a `verify_credentials` method.

**Reply by arnonrdp** [reply]
My first attempt was to continue with the "old" code, but when it tries to compose the tweet, I get this error:
> Erro:  403 Forbidden
453 - You currently have access to a subset of Twitter API v2 endpoints and limited v1.1 endpoints (e.g. media post, oauth) only. If you need access to this endpoint, you may need a different access level.

This is what I used to do:
```py
from os import getenv
from time import sleep
from tweepy import OAuthHandler, API

def twitter_connect(tweet_list):
    auth = OAuthHandler(getenv('API_KEY'),
                        getenv('API_KEY_SECRET'))
    auth.set_access_token(getenv('ACCESS_TOKEN'),
                          getenv('ACCESS_TOKEN_SECRET'))
    api = API(auth)

    try:
        api.verify_credentials()
        print('Good news, you are logged!')
        prepare_tweet(tweet_list, api)
    except Exception as e:
        print('Bad news, will try again in 2 minutes.\n', e)
        sleep(120)
        twitter_connect(tweet_list)
```

This doesn't make sense to me since my App uses v2 access:

<img width="1118" alt="Captura de Tela 2023-07-07 às 05 22 30" src="https://github.com/tweepy/tweepy/assets/26576201/fa2e9064-dc01-4194-bbd1-49e16b2860d3">

**Reply by Harmon758** [reply]
`tweepy.API` is Tweepy's interface for Twitter API v1.1.
https://tweepy.readthedocs.io/en/v4.14.0/api.html#tweepy.API

**Reply by arnonrdp** [reply]
Yes,  but according to you:
> You can continue using OAuth 1.0a User Context with Twitter API v2 though.

If I need to create tweets, how do I resolve this error?
> 453 - You currently have access to a subset of Twitter API v2 endpoints and limited v1.1 endpoints (e.g. media post, oauth) only. If you need access to this endpoint, you may need a different access level.

**Reply by Harmon758** [reply]
Yes.
Use Twitter API v2.

**Reply by arnonrdp** [reply]
I got your point.
You told me to use the authorization URL, instead of just passing it to `OAuth2UserHandler.fetch_token`.

But how am I supposed to login/authorize in v2 if this is running in a GitHub action?

**Reply by Harmon758** [reply]
Again, as I said in my original answer, you can continue using OAuth 1.0a User Context with Twitter API v2.

**Reply by arnonrdp** [reply]
And again, I tell you: this is returning error 453.

This is a infinite loop.

**Reply by Harmon758** [reply]
You need to provide the full traceback and the relevant code.

**Reply by arnonrdp** [reply]
I did.
My questions are:
1. For an application launched via a Github Action, is my callback correct?
2. How do I perform authentication through a Github Action?

**Reply by Harmon758** [reply]
The only full traceback you've provided is for a `oauthlib.oauth2.rfc6749.errors.MissingCodeError` involving OAuth 2.0 Authorization Code Flow with PKCE, which my original answer addresses.

You've provided no code that is using OAuth 1.0a User Context with Twitter API v2.

You've also provided no code that is using a loop of any kind, so I'm not sure how you'd be encountering an infinite loop, unless you're using a process manager or something, which would be well outside the scope of the library.

So no, you have not provided the full traceback and the relevant code for any "error 453" or infinite loop you're encountering from using OAuth 1.0a User Context with Twitter API v2.

As for these two questions, I'm not sure what callback you're referring to, and again, you can continue using OAuth 1.0a User Context with Twitter API v2. I'm not going to keep answering the same question over and over again.

**Reply by arnonrdp** [reply]
The loop I mention is about our conversation lol.

Here is the code and Traceback:

![Captura de Tela 2023-07-07 às 19 48 27](https://github.com/tweepy/tweepy/assets/26576201/d8112368-b4b9-45fc-8eef-e0d5401b9b27)

Just to summarize/explain: The application takes a newsletter, turns it into a list, and posts each list item as a tweet.

---
EDIT:

I know line 13 is wrong, but I am stuck in here, I don't know how to proceed with this authentication.

**Reply by Harmon758** [reply]
If you insist on using OAuth 2.0 Authorization Code Flow with PKCE with GitHub Actions, you'll need to use a refresh token.
https://developer.twitter.com/en/docs/authentication/oauth-2-0/authorization-code

I'm not sure why you would do this instead of just continuing to use OAuth 1.0a User Context though, if you're not using any endpoints that require OAuth 2.0 Authorization Code Flow with PKCE.

**Reply by arnonrdp** [reply]
Because I think OAuth 1.0 is not working to me, it says:
> 453 - You currently have access to a subset of Twitter API v2 endpoints and limited v1.1 endpoints (e.g. media post, oauth) only. If you need access to this endpoint, you may need a different access level.

The only thing my app does is **post tweets**.

**Reply by Harmon758** [reply]
Again, you need to provide to provide the full traceback and the relevant code.
I'm not going to keep telling you this.

**Comment by leonardo-cordeiro** [comment, marked as answer]
Try this:
```py
from tweepy import Client


def twitter_connect(tweet_list):
    client = Client(
        consumer_key=getenv('API_KEY'),
        consumer_secret=getenv('API_KEY_SECRET'),
        access_token=getenv('ACCESS_TOKEN'),
        access_token_secret=getenv('ACCESS_TOKEN_SECRET')
    )

    prepare_tweet(tweet_list, client)
```

**Reply by arnonrdp** [reply]
That what I was looking for!

Worked perfectly, thanks bro.

**Reply by Harmon758** [reply]
I've been telling you to do this for two days now, including in my original answer, while you repeatedly kept saying this errored for you, without providing the full traceback or relevant code. I'm not sure if this was to purposefully waste my time or antagonize me, but I certainly don't appreciate it.
